mongo/client_bulk_write.go

```diff
@@ -0,0 +1,703 @@
+	namespace int
+	document  interface{}
+}
+
+func (d *clientInsertDoc) marshal(bsonOpts *options.BSONOptions, registry *bson.Registry) (interface{}, bsoncore.Document, error) {
+	uidx, doc := bsoncore.AppendDocumentStart(nil)
+
+	doc = bsoncore.AppendInt32Element(doc, "insert", int32(d.namespace))
+	f, err := marshal(d.document, bsonOpts, registry)
+	if err != nil {
+		return nil, nil, err
+	}
+	var id interface{}
+	f, id, err = ensureID(f, bson.NilObjectID, bsonOpts, registry)
+	if err != nil {
+		return nil, nil, err
+	}
+	doc = bsoncore.AppendDocumentElement(doc, "document", f)
+	doc, err = bsoncore.AppendDocumentEnd(doc, uidx)
+	return id, doc, err
+}
+
+type clientUpdateDoc struct {
+	namespace      int
+	filter         interface{}
+	update         interface{}
+	hint           interface{}
+	arrayFilters   []interface{}
+	collation      *options.Collation
+	upsert         *bool
+	multi          bool
+	checkDollarKey bool
+}
+
+func (d *clientUpdateDoc) marshal(bsonOpts *options.BSONOptions, registry *bson.Registry) (bsoncore.Document, error) {
+	uidx, doc := bsoncore.AppendDocumentStart(nil)
+
+	doc = bsoncore.AppendInt32Element(doc, "update", int32(d.namespace))
+
+	f, err := marshal(d.filter, bsonOpts, registry)
+	if err != nil {
+		if d.filter == nil {
+			return nil, fmt.Errorf("%w: filter is required", err)
+		}
+		return nil, err
+	}
+	doc = bsoncore.AppendDocumentElement(doc, "filter", f)
+
+	u, err := marshalUpdateValue(d.update, bsonOpts, registry, d.checkDollarKey)
+	if err != nil {
+		return nil, err
+	}
+	doc = bsoncore.AppendValueElement(doc, "updateMods", u)
+	doc = bsoncore.AppendBooleanElement(doc, "multi", d.multi)
+
+	if d.arrayFilters != nil {
+		reg := registry
+		arr, err := marshalValue(d.arrayFilters, bsonOpts, reg)
+		if err != nil {
+			return nil, err
+		}
+		doc = bsoncore.AppendArrayElement(doc, "arrayFilters", arr.Data)
+	}
+
+	if d.collation != nil {
+		doc = bsoncore.AppendDocumentElement(doc, "collation", toDocument(d.collation))
+	}
+
+	if d.upsert != nil {
+		doc = bsoncore.AppendBooleanElement(doc, "upsert", *d.upsert)
+	}
+
+	if d.hint != nil {
+		if isUnorderedMap(d.hint) {
+			return nil, ErrMapForOrderedArgument{"hint"}
+		}
+		hintVal, err := marshalValue(d.hint, bsonOpts, registry)
+		if err != nil {
+			return nil, err
+		}
+		doc = bsoncore.AppendValueElement(doc, "hint", hintVal)
+	}
+
+	return bsoncore.AppendDocumentEnd(doc, uidx)
+}
+
+type clientDeleteDoc struct {
+	namespace int
+	filter    interface{}
+	collation *options.Collation
+	hint      interface{}
+	multi     bool
+}
+
+func (d *clientDeleteDoc) marshal(bsonOpts *options.BSONOptions, registry *bson.Registry) (bsoncore.Document, error) {
+	didx, doc := bsoncore.AppendDocumentStart(nil)
+
+	doc = bsoncore.AppendInt32Element(doc, "delete", int32(d.namespace))
+
+	f, err := marshal(d.filter, bsonOpts, registry)
+	if err != nil {
+		if d.filter == nil {
+			return nil, fmt.Errorf("%w: filter is required", err)
+		}
```

Comment:
Is there a reason to nest this nil filter check in this error handling block? Can we move this check to the beginning of `marshal`?

Optional: Consider a clearer error message, like
> delete filter cannot be nil